### ISSUE TITLE
resume tokens are only partially checksummed

### ISSUE THREAD

**Issue by cfcs** [openzfs/zfs#10517]
### System information
Type | info
 --- | ---
Distribution Name	| Fedora
Distribution Version	| not sure
Linux Kernel	| 4.19
Architecture	| amd64
ZFS Version	| 0.8.0-1
SPL Version	| 0.8.0-1

### Describe the problem you're observing
When emitting a resume token, the token is constructed to consist of:
- a (token format) version number;
- a (truncated) fletcher 4 checksum ;
- the length of the decompressed payload ;
- gzip-compressed payload.

This is implemented in [dsl_dataset.c:get_receive_resume_stats_impl](https://github.com/openzfs/zfs/blob/5a42ef04fd390dc96fbbf31bc9f3d05695998211/module/zfs/dsl_dataset.c#L2428-L2442).

The function that decodes these values is [libzfs_sendrecv.c:zfs_send_resume_token_to_nvlist()](https://github.com/openzfs/zfs/blob/ae7b167a98b48260554fd88cef022547283e125f/lib/libzfs/libzfs_sendrecv.c#L1347)

The fletcher4 checksum is generated by calling [zfs_fletcher.c:fletcher_4_native_varsize()](https://github.com/openzfs/zfs/blob/b3673342c79b65c441c98a8161ffbf136d497c2f/module/zcommon/zfs_fletcher.c#L488) with the pointer to and length of the gzip-compressed payload.

Presumably since fletcher_4_native is designed to operate on nvlist/nvpair data, there exists an invariant that the to-be-checksummed [buffer is 32-bit aligned](https://github.com/openzfs/zfs/blob/b3673342c79b65c441c98a8161ffbf136d497c2f/module/zcommon/zfs_fletcher.c#L321) (and the 0-3 extra bytes are silently ignored).

Since the resume token decided to take a detour from this design pattern, and since the gzip compression function only emits 32-bit aligned values 25% of the time, the invariant that the argument to fletcher_4_native is word-aligned is often violated. 

This results in `zfs receive -s` frequently emitting tokens that are only partially protected by the fletcher4 checksum.

So let's say I had accidentally managed to add/delete a few bytes at the end when copy-pasting, would this result in ZFS doing the wrong thing? I don't know, I do not have sufficient insight into the encoding rules of gzip, `nvlist`s, and/or the internal workings of ZFS. It seems plausible to me that there could exist some last-few-character manipulations that would result in different, same-length decompressed payloads, but I don't know.
Anyway, silently ignoring data when checksumming also seems like a quirk you'd want to have a `assert()` or at the very least a scary log message about for the developers?

Personal feedback:
1) Personally, I was motivated to raise this issue after having spent an hour jumping through ZFS code to figure out why the hell my own implementation of fletcher4 would "often" disagree with `zfs send` about what constituted the correct checksum for a given payload. I've been writing a small "firewall" for resume tokens that aims to make sure that the remote backup server cannot retrieve data from the sending server besides the diff between the two original snapshots/bookmarks. The fact that resume tokens should essentially be considered *untrusted input* (in terms of information security) was not clear to me before reading the code, so perhaps that could be featured more **prominently** in the documentation.
2) If you decide to write a "version 2" of the resume token format at some point, I think it would be amazing if you would consider validating the nvlist dict against the command-line arguments given to `zfs send` to ensure it doesn't stray completely off to other datasets or asks for terabytes of data when the sender meant to send a 1MB diff. Or maybe there could be a send-side HMAC to bound the requested object set. But all that is probably a different discussion.
3) Truncating the fletcher4 checksum by only using the first zcword severely degrades the usefulness of the checksum since all words are now interchangeable, and it is unable to detect extraneous zeros. CRC-32 or some other more robust scheme would seem worth considering as well if a new format is developed.

### Describe how to reproduce the problem

The problem will manifestate by itself, the way I generated the token below was to run something along the lines of:
```shell
zfs create zroot/empty
zfs snapshot zroot/empty@a
zfs snapshot zroot/empty@b
zfs send zroot/empty@a | zfs receive zroot/empty4
zfs send -i zroot/empty@a zroot/empty@b > file
head -c 400 file | zfs receive -s zroot/empty4
zfs get resumetok zroot/empty4
```

If you take the last part of the resume token, divide the length by two (since it's hex-encoded), and modulo 4 of this is not zero, you have a "bad" token (it will work perfectly fine, just the checksum doesn't cover the last few bytes).

Here is one such token:
```
1-b7e234b96-e0-789c636064000310a501c49c50360710a715e5e7a69766a63040c1b3135ff83ff4cb3b2a00d9ec48eaf293b252934b207c10c0904f4b2b4e2d618003903c1b927c5265496a3103aa3cb2fe927c882b5463e3fb24ce15ab6b20c97382e5f31273531918aa8af2f34bf453730b4a2a1d608e060062051d87
```

If you try to resume a transfer from your pool using this token, it will eventually fail with an error about how your datasets are not the same as mine, so it may be more instructive to first generate your own token following the procedure outlined above before proceding with the next steps.

I've replaced the token with $TOKEN in the steps below to make it more readable.

- Anyway, first we try to resume the transfer (all good, we get the binary replication stream):
```shell
zfs send -t $TOKEN | cat
`_@G^%]_s'Azroot/empty@dresume_objectresume_offset
```

- Then we try adding AAAA to the end ([relevant code](https://github.com/openzfs/zfs/blob/ae7b167a98b48260554fd88cef022547283e125f/lib/libzfs/libzfs_sendrecv.c#L1394)) - so far so good, this is expected:
```shell
zfs send -t "$TOKEN"AAAA
cannot resume send: resume token is corrupt (incorrect checksum)
```

- Next we try altering the last byte (I incremented my 87 to 88) ([relevant code](https://github.com/openzfs/zfs/blob/ae7b167a98b48260554fd88cef022547283e125f/lib/libzfs/libzfs_sendrecv.c#L1406)) - notice how it passes the checksum test and now fails because I failed to terminate my gzip stream correctly:
```shell
zfs send -t $TOKEN | cat
cannot resume send: resume token is corrupt (decompression failed)
```

### Include any warning/errors/backtraces from the system logs

Here is a dtrace one-liner to demonstrate the problem (observe that the size is 119 ; not a multiple of 4):
```
dtrace -n 'pid$target::fletcher_4_native:entry {printf("size of buf: %u\n",arg1);ustack();}' -c 'zfs send -t 1-b7e234b96-e0-789c636064000310a501c49c50360710a715e5e7a69766a63040c1b3135ff83ff4cb3b2a00d9ec48eaf293b252934b207c10c0904f4b2b4e2d618003903c1b927c5265496a3103aa3cb2fe927c882b5463e3fb24ce15ab6b20c97382e5f31273531918aa8af2f34bf453730b4a2a1d608e060062051d87'|cat

CPU     ID                    FUNCTION:NAME
  1  73197          fletcher_4_native:entry size of buf: 119

              libzfs.so.3`fletcher_4_native+0x1
              libzfs.so.3`zfs_send_resume_token_to_nvlist+0xef
              libzfs.so.3`zfs_send_resume+0x77
              zfs`zfs_do_send+0x2ee
              zfs`main+0x1ef
              zfs`_start+0x10d
              zfs`0x10b516a05008

```

**Comment by cfcs** [comment]
Err, the `dtrace` came from a freebsd system, but same thing there.

**Comment by cfcs** [comment]
Oh, perhaps also worth nothing: My implementation of incremental fletcher4 did not do the 32-bit unrolling, but instead kept track of the state (byte shift `0 | 1 | 2 |3`) and operated byte-wise. Not sure if that's practical for ZFS, but it could be an option if padding the compressed data is not an option.

**Comment by misterbigstuff** [comment]
ZFS checksums are not meant to be cryptographically secure, and AIUI, ZFS encryption also prevents people from reading data but not modifying it. @tcaputi stated this during encryption development.

**Comment by cfcs** [comment]
@misterbigstuff TL;DR of the problem I attempted to describe here is that in some instances you have data like resume token `ABCDEFGHijk`, and the checksum for the resume tokens `ABCDEFGH`, `ABCDEFGHijk`, and `ABCDEFGHxyz` all turn out to be identical (because it doesn't try to calculate the checksum for the last few letters), which means you can accidentally end up with `ABCDEFGH` if you botch a copy-paste or whatever.

A cryptographic checksum would have the same problem: If you have 10 bytes and you ask for a SHA-3 hash of the first 8 bytes, your cryptographic checksum doesn't cover the remaining two bytes. It's probably not the end of the world here, but I guess it's something to watch out for in other cases where ZFS tries to make use of checksums.

About choice of checksum: The way checksumming is done here means that furthermore `EFGHABCDxyz` and `ABCDEFGHmno` also have the same checksum. Fletcher's algorithm was developed in response to the problem of block reordering, and truncating it to the first word the way it's done here makes that property come undone, reducing the call to fletcher_4_native to a slightly expensive version of `uint32_t* p, endp; uint32_t cs=0; while (p != endp) {cs += *p; p++;}`. Again not the end of the world in this particular instance, but not so elegant.

**Comment by cfcs** [comment]
I don't think it should be closed, particularly because this impartial checksum might accidentally be used elsewhere, too, where it has a bigger impact on correctness.

**Comment by ryao** [comment]
> I don't think it should be closed, particularly because this impartial checksum might accidentally be used elsewhere, too, where it has a bigger impact on correctness.

A bot did that. I just reopened it manually.
